epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyTouchHelperCallback.kt

```diff
@@ -0,0 +1,222 @@
+package com.airbnb.epoxy
+
+import android.graphics.Canvas
+import androidx.recyclerview.widget.ItemTouchHelper
+import androidx.recyclerview.widget.RecyclerView
+
+/**
+ * A wrapper around [androidx.recyclerview.widget.ItemTouchHelper.Callback] to cast all
+ * view holders to [com.airbnb.epoxy.EpoxyViewHolder] for simpler use with Epoxy.
+ */
+abstract class EpoxyTouchHelperCallback : ItemTouchHelper.Callback() {
+
+    override fun getMovementFlags(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder) =
+        getMovementFlags(recyclerView, viewHolder as EpoxyViewHolder)
+
+    /**
+     * @see getMovementFlags
+     */
+    protected abstract fun getMovementFlags(
+        recyclerView: RecyclerView,
+        viewHolder: EpoxyViewHolder
+    ): Int
+
+    override fun onMove(
+        recyclerView: RecyclerView,
+        viewHolder: RecyclerView.ViewHolder,
+        target: RecyclerView.ViewHolder
+    ) =
+        onMove(recyclerView, viewHolder as EpoxyViewHolder, target as EpoxyViewHolder)
+
+    /**
+     * @see onMove
+     */
+    protected abstract fun onMove(
+        recyclerView: RecyclerView,
+        viewHolder: EpoxyViewHolder,
+        target: EpoxyViewHolder
+    ): Boolean
+
+    override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) =
+        onSwiped(viewHolder as EpoxyViewHolder, direction)
+
+    /**
+     * @see onSwiped
+     */
+    protected abstract fun onSwiped(viewHolder: EpoxyViewHolder, direction: Int)
+
+    override fun canDropOver(
+        recyclerView: RecyclerView,
+        current: RecyclerView.ViewHolder,
+        target: RecyclerView.ViewHolder
+    ) =
+        canDropOver(recyclerView, current as EpoxyViewHolder, target as EpoxyViewHolder)
+
+    /**
+     * @see canDropOver
+     */
+    protected open fun canDropOver(
+        recyclerView: RecyclerView,
+        current: EpoxyViewHolder,
+        target: EpoxyViewHolder
+    ) =
+        super.canDropOver(recyclerView, current, target)
+
+    override fun getSwipeThreshold(viewHolder: RecyclerView.ViewHolder) =
```

Comment:
can you add explicit return types for all of these functions? best to do that for public library interfaces

Reply:
sure, done